dbms/src/Functions/FunctionsJson.h

```diff
@@ -530,11 +612,671 @@ class FunctionsJsonArray : public IFunction
+        JsonBinary::JsonBinaryWriteBuffer write_buffer(data_to, reserve_size);
+
+        for (size_t i = 0; i < data_from.size(); ++i)
+        {
+            JsonBinary::appendNumber(write_buffer, static_cast<ToType>(data_from[i]));
+            writeChar(0, write_buffer);
+            offsets_to[i] = write_buffer.count();
+        }
+    }
+
+private:
+    const tipb::FieldType * input_tidb_tp = nullptr;
+};
+
+class FunctionCastStringAsJson : public IFunction
+{
+public:
+    static constexpr auto name = "cast_string_as_json";
+    static FunctionPtr create(const Context &) { return std::make_shared<FunctionCastStringAsJson>(); }
+
+    String getName() const override { return name; }
+
+    size_t getNumberOfArguments() const override { return 1; }
+
+    bool useDefaultImplementationForNulls() const override { return false; }
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    void setInputTiDBFieldType(const tipb::FieldType & tidb_tp_) { input_tidb_tp = &tidb_tp_; }
+    void setOutputTiDBFieldType(const tipb::FieldType & tidb_tp_) { output_tidb_tp = &tidb_tp_; }
+    void setCollator(const TiDB::TiDBCollatorPtr & collator_) override { collator = collator_; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        const auto & input_type = arguments[0];
+        if (input_type->onlyNull())
+        {
+            return input_type;
+        }
+
+        if unlikely (!removeNullable(input_type)->isStringOrFixedString())
+            throw Exception(
+                fmt::format("Illegal type {} of argument of function {}", arguments[0]->getName(), getName()),
+                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+        auto return_type = std::make_shared<DataTypeString>();
+        return input_type->isNullable() ? makeNullable(return_type) : return_type;
+    }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result) const override
+    {
+        const auto & from = block.getByPosition(arguments[0]);
+        if (from.type->onlyNull())
+        {
+            auto only_null_column = from.column;
+            block.getByPosition(result).column = std::move(only_null_column);
+            return;
+        }
+
+        auto nested_block = createBlockWithNestedColumns(block, arguments);
+        auto input_source = createDynamicStringSource(*nested_block.getByPosition(arguments[0]).column);
+
+        auto col_to = ColumnString::create();
+        auto & data_to = col_to->getChars();
+        auto & offsets_to = col_to->getOffsets();
+        auto rows = block.rows();
+        offsets_to.resize(rows);
+
+        // In raw function test, input_tidb_tp/output_tidb_tp is nullptr.
+        if (collator && collator->isBinary())
+        {
+            if (unlikely(input_tidb_tp == nullptr))
+            {
+                doExecuteForBinary<false>(data_to, offsets_to, input_source, input_tidb_tp->tp(), -1, block.rows());
```

Comment:
will it be an invalid pointer access if call `input_tidb_tp->tp()` when `input_tidb_tp` is nullptr?

Reply:
yes....my mistake, updated.